# Functions

### Let's solve second degree equations
$$ ax^2 + bx + c = 0 $$
if $a = 0$
$$ x_1=x_2=-\frac{c}{b} $$
if $a \neq 0$

$$ \Delta = b^2 - 4ac $$

$$ x_{1,2} = \frac{-b \pm \sqrt{\Delta}}{2a}$$

In [1]:
def second_degree_eq_solver(a,b,c):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")

    if a == 0:
        x1 = -c/b
        print("The solution is:",x1)
    else:
        x1 = (-b-(b**2-4*a*c)**0.5)/(2*a)
        x2 = (-b+(b**2-4*a*c)**0.5)/(2*a)
        print("The solutions are: ",x1,x2)

second_degree_eq_solver(1., 3., 2.)

solving  1.0 x^2 +  3.0 x +  2.0 = 0
The solutions are:  -2.0 -1.0


* `def` defines a function
* to invoke the function, after the function name the argument list has to be passed
* the function executes its own statements and then exits to the calling unit
* passed variable (actual arguments) are matched to local variables (dummy arguments)

### Returning from a function

In [2]:
def second_degree_eq_solver(a,b,c):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")

    if a == 0:
        x1 = -c/b
        number_of_solutions = 1
    else:
        x1 = (-b-(b**2-4*a*c)**0.5)/(2*a)
        x2 = (-b+(b**2-4*a*c)**0.5)/(2*a)
        number_of_solutions = 2
    
    return number_of_solutions
    
n_solutions = second_degree_eq_solver(1., 3., 2.)
print("number of solutions: ", n_solutions)

solving  1.0 x^2 +  3.0 x +  2.0 = 0
number of solutions:  2


* A function may `return` a value
   * If no value is returned, implicitly `None` is returned

In [3]:
def second_degree_eq_solver(a,b,c):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")

    x1 = x2 = None
    if a == 0:
        x1 = -c/b
    else:
        x1 = (-b-(b**2-4*a*c)**0.5)/(2*a)
        x2 = (-b+(b**2-4*a*c)**0.5)/(2*a)
    return x1, x2   # <- multiple values returned
    
sol_1, sol_2 = second_degree_eq_solver(0., 3., 2.)
print("solutions are: ", sol_1, sol_2)

solving  0.0 x^2 +  3.0 x +  2.0 = 0
solutions are:  -0.6666666666666666 None


* A function may also return multiple values just separating them by commas

### Calling function from another function

In [4]:
def delta(a,b,c):
    return b**2-4*a*c

def second_degree_eq_solver(a,b,c):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")

    x1 = x2 = None
    if a == 0:
        x1 = -c/b
    else:
        x1 = (-b-delta(a,b,c)**0.5)/(2*a)
        x2 = (-b+delta(a,b,c)**0.5)/(2*a)
    return x1, x2
    
print("test delta: ", delta(1, 3, 2))
sol_1, sol_2 = second_degree_eq_solver(1., 3., 2.)
print("solutions are: ", sol_1, sol_2)

test delta:  1
solving  1.0 x^2 +  3.0 x +  2.0 = 0
solutions are:  -2.0 -1.0


* A function may call another function
   * a good practice is usually to write many functions to split the tasks of the code, improving readability, reducing code replication

### Functions inside other functions

In [5]:
def second_degree_eq_solver(a,b,c):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")

    def delta_inside(a,b,c):
        return b**2-4*a*c
    
    x1 = x2 = None
    if a == 0:
        x1 = -c/b
    else:
        x1 = (-b-delta_inside(a,b,c)**0.5)/(2*a)
        x2 = (-b+delta_inside(a,b,c)**0.5)/(2*a)
    return x1, x2
    
sol_1, sol_2 = second_degree_eq_solver(1., 3., 2.)
print("solutions are: ", sol_1, sol_2)

solving  1.0 x^2 +  3.0 x +  2.0 = 0
solutions are:  -2.0 -1.0


In [6]:
print("delta: ", delta(1, 2, 3))

delta:  -8


In [7]:
print("delta inside: ", delta_inside(1, 2, 3))

NameError: name 'delta_inside' is not defined

* Functions can be nested
   * improves readability
   * improves safety, functions can be called only in the units defining them
   * but makes separate tests harder

### Trying to get solutions from arguments (sort of _pass by reference_)

In [8]:
def second_degree_eq_solver(a,b,c,x1,x2):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")

    def delta_inside(a,b,c):
        return b**2-4*a*c
    
    if a == 0:
        x1 = -c/b
    else:
        x1 = (-b-delta(a,b,c)**0.5)/(2*a)
        x2 = (-b+delta(a,b,c)**0.5)/(2*a)
    
sol_1 = sol_2 = None
second_degree_eq_solver(0., 3., 2., sol_1, sol_2)
print("solutions are: ", sol_1, sol_2) # why is this not working ?

solving  0.0 x^2 +  3.0 x +  2.0 = 0
solutions are:  None None


* We could think to pass the solutions of equations as additional arguments instead of returning them
* It does not work. Why?
   * Because reassigning scalars inside the function creates new variables (i.e. label to a different object) so that, when exiting the function, the passed variables still have the old values
   * Not possible to fix it for scalars, they are `immutable`!

### Default arguments

In [9]:
def second_degree_eq_solver(a=0., b=0., c=0.):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")
    
    def delta_inside(a,b,c):
        return b**2-4*a*c
    
    x1 = x2 = None
    if a == 0:
        x1 = -c/b
    else:
        x1 = (-b-delta_inside(a,b,c)**0.5)/(2*a)
        x2 = (-b+delta_inside(a,b,c)**0.5)/(2*a)
    return x1, x2

sol_1, sol_2 = second_degree_eq_solver(-1., 4.)
# sol_1, sol_2 = second_degree_eq_solver(c=-1., a=4.)
# sol_1, sol_2 = second_degree_eq_solver(-1., c=2.)

print("solutions are: ", sol_1, sol_2)

#### wrong argument calls
# sol_1, sol_2 = second_degree_eq_solver(c=-1., a=4., 3.) # raises SyntaxError
# sol_1, sol_2 = second_degree_eq_solver(-1., 4., a=4.)# raises TypeError

solving  -1.0 x^2 +  4.0 x +  0.0 = 0
solutions are:  4.0 -0.0


* Arguments may have default values
   * not passed arguments retain the default values inside the function
* And arguments can be passed using keywords
   * improves readability
   * order of arguments can be altered
* non-keywords argument go first, other takes defaults
   * beware of order and duplicates

### Sketching a workflow. The `pass` statement

In [34]:
def check_results(s1, s2):
    pass

def second_degree_eq_solver(a=0., b=0., c=0.):
    print("solving ", a, "x^2 + ", b, "x + ", c, "= 0")

    def delta_inside(a,b,c):
        return b**2-4*a*c
    
    x1 = x2 = None
    if a == 0:
        x1 = -c/b
    else:
        x1 = (-b-delta_inside(a,b,c)**0.5)/(2*a)
        x2 = (-b+delta_inside(a,b,c)**0.5)/(2*a)
    return x1, x2
    
sol_1, sol_2 = second_degree_eq_solver(c=-1., a=4.)

check_results(sol_1, sol_2)

print("solutions are: ", sol_1, sol_2)

solving  4.0 x^2 +  0.0 x +  -1.0 = 0
solutions are:  -0.5 0.5


* You can express a workflow prototyping actions which you'll implement later
* Each function must be found when calling it
* And cannot be empty:
    * use 'pass' to skip implementation

### Use functions to compute trapezoidal integration of $f$ between $[a,b]$


In [11]:
def trapezoidal_integration(a, b, tol, f, maxsteps=2**23):
    # initialization
    intervals = 1 ; acc = (f(a) + f(b))*0.5 ; dx = (b - a) / intervals
    estimate = acc * dx ; prev_estimate = estimate + 2*tol 
    convergence = False ; steps = 1
    # iterative estimate
    while abs(estimate - prev_estimate) > tol:
        print("Current estimate using ", intervals, " intervals with dx =", dx, " is ", estimate)
        prev_estimate = estimate
        intervals *= 2 ; dx *= 0.5
        for i in range(1, steps + 1, 2): # only contributions from new points
            x = a + i*dx ; acc += f(x)
        estimate = acc*dx
        steps = steps + 1
        if steps > maxsteps: 
            break
    else:
        convergence=True
    return estimate, convergence

def integrand(x):
    return x**2

* It is possible to (easily) **pass functions as arguments**
   * our integral function becomes very general
* Selecting functions at run-time is very easy (very powerful feature)

In [12]:
integral, success = trapezoidal_integration(a=-1.0, b=2.0, tol=1.e-4, f=integrand)
print("integral, success: ", integral, success)

Current estimate using  1  intervals with dx = 3.0  is  7.5
Current estimate using  2  intervals with dx = 1.5  is  4.125
Current estimate using  4  intervals with dx = 0.75  is  2.109375
Current estimate using  8  intervals with dx = 0.375  is  1.20703125
Current estimate using  16  intervals with dx = 0.1875  is  0.76318359375
Current estimate using  32  intervals with dx = 0.09375  is  0.533477783203125
Current estimate using  64  intervals with dx = 0.046875  is  0.3714179992675781
Current estimate using  128  intervals with dx = 0.0234375  is  0.26296234130859375
Current estimate using  256  intervals with dx = 0.01171875  is  0.17409682273864746
Current estimate using  512  intervals with dx = 0.005859375  is  0.11466186493635178
Current estimate using  1024  intervals with dx = 0.0029296875  is  0.07155436556786299
Current estimate using  2048  intervals with dx = 0.00146484375  is  0.04441264900378883
Current estimate using  4096  intervals with dx = 0.000732421875  is  0.02656

### Lambdas

* Simple functions can be defined when needed with one-liner lambda expressions
* `lambda x: x**2` defines the parabola function 
  * can reference variables from the containing scope
  * syntactically restricted to a single expression
  * just syntactic sugar for a normal function definition

In [13]:
my_func = lambda x: x**2
integral, success = trapezoidal_integration(a=-1.0, b=2.0, tol=1.e-4, f=my_func)
print("integral, success: ", integral, success)

Current estimate using  1  intervals with dx = 3.0  is  7.5
Current estimate using  2  intervals with dx = 1.5  is  4.125
Current estimate using  4  intervals with dx = 0.75  is  2.109375
Current estimate using  8  intervals with dx = 0.375  is  1.20703125
Current estimate using  16  intervals with dx = 0.1875  is  0.76318359375
Current estimate using  32  intervals with dx = 0.09375  is  0.533477783203125
Current estimate using  64  intervals with dx = 0.046875  is  0.3714179992675781
Current estimate using  128  intervals with dx = 0.0234375  is  0.26296234130859375
Current estimate using  256  intervals with dx = 0.01171875  is  0.17409682273864746
Current estimate using  512  intervals with dx = 0.005859375  is  0.11466186493635178
Current estimate using  1024  intervals with dx = 0.0029296875  is  0.07155436556786299
Current estimate using  2048  intervals with dx = 0.00146484375  is  0.04441264900378883
Current estimate using  4096  intervals with dx = 0.000732421875  is  0.02656

### Recursion

In [14]:
def factorial(n):
    if n < 2:
        return 1
    else:
        return n * factorial(n - 1)
    
factorial(18)

6402373705728000

### Scopes
* Each function defines a local scope
* Variables outside any function are globally scoped and can be accessed inside the nested functions
   * global variables can easily make your code less readable/maintainable!

In [15]:
a = 4

def do_something():
    print("inside a: ", a)
    
print("outside before a: ",a)
do_something()
print("outside after a: ", a)

outside before a:  4
inside a:  4
outside after a:  4


In [16]:
a = 4
def do_something():
    print("inside a: ", a)
    
print("outside before a: ", a)
a = 2
do_something()
print("outside after a: ", a)

outside before a:  4
inside a:  2
outside after a:  2


* If `a` is reassigned in the function scope, it is a new object and it becomes the only `a` known in that scope

In [17]:
a = 4

def do_something():
    a=2
    print("inside after a: ", a)
    
print("outside before a: ", a)
do_something()
print("outside after a: ", a)

outside before a:  4
inside after a:  2
outside after a:  4


* If printing it before assigning it, an error will arise

In [18]:
a = 4

def do_something():
    print("inside before a: ", a)
    a=2
    print("inside after a: ", a)
    
print("outside before a: ", a)
do_something()
print("outside after a: ", a)

outside before a:  4


UnboundLocalError: cannot access local variable 'a' where it is not associated with a value

* unless `global a` is declared in the function (before using it) to clarify that `a` must have global scope

In [19]:
a = 4

def do_something():
    global a
    print("inside before a: ",a)
    a = 2
    print("inside after a: ",a)
    
print("outside before a: ",a)
do_something()
print("outside after a: ",a)

outside before a:  4
inside before a:  4
inside after a:  2
outside after a:  2


* symptom of bad design...  avoid it, change names or pass variables as arguments


In [20]:
a = 4

def do_something(a):
    print("inside before a: ", a)
    a = 2
    print("inside after a: ", a)
    
print("outside before a: ", a)
do_something(a)
print("outside after a: ", a)

outside before a:  4
inside before a:  4
inside after a:  2
outside after a:  4


* when the number of arguments becomes large, Python has suitable features ...

### Documenting functions
* Documenting functions is often a good idea
* The Python way is adding a multiline string just after the `def` line (a [**docstring**](https://peps.python.org/pep-0257/))
   * a multiline string in Python is surrounded by triple quotes `"""`
* Following a clean and uniform style allows to easily find what is needed
* The function docs can be accessed externally without browsing the function code (see later)

In [21]:
def second_degree_eq_solver(a=0., b=0., c=0.):
    """Function which solves the second degree equation.
    
    Args:
        a (float or complex, optional): second degree coefficient. Defaults to zero
        b (float or complex, optional): first degree coefficient. Defaults to zero
        c (float or complex, optional): zero degree coefficient. Defaults to zero
        
    Returns:
        x1, x2 (float or complex): the solutions of the equation
        
        For equations having only one solution x2 is returned as None
    """
    
    # HERE GOES THE IMPLEMENTATION....
    
    return x1, x2

### Documentation String Rules
* first line should always be a short, concise summary of the object’s purpose. 
  * should begin with a capital letter and end with a period.
* if there are more lines in the documentation string, the second line should be blank
* other lines should describe the object’s calling conventions, its side effects, etc.

In [22]:
help(second_degree_eq_solver)

Help on function second_degree_eq_solver in module __main__:

second_degree_eq_solver(a=0.0, b=0.0, c=0.0)
    Function which solves the second degree equation.

    Args:
        a (float or complex, optional): second degree coefficient. Defaults to zero
        b (float or complex, optional): first degree coefficient. Defaults to zero
        c (float or complex, optional): zero degree coefficient. Defaults to zero

    Returns:
        x1, x2 (float or complex): the solutions of the equation

        For equations having only one solution x2 is returned as None



### Docstrings styles: Google vs NumPy
There are two styles of docstrings: [Google](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings) and [NumPy](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard). The main difference between the two styles is that Google uses indentation to separate sections, whereas NumPy uses underlines.

#### Google style:

```python
def func(arg1, arg2):
    """Summary line.

    Extended description of function.

    Args:
        arg1 (int): Description of arg1
        arg2 (str): Description of arg2

    Returns:
        bool: Description of return value

    """
    return True
```
#### NumPy style:

```python
def func(arg1, arg2):
    """Summary line.

    Extended description of function.

    Parameters
    ----------
    arg1 : int
        Description of arg1
    arg2 : str
        Description of arg2

    Returns
    -------
    bool
        Description of return value

    """
    return True
```
Google style tends to be easier to read for short and simple docstrings, whereas NumPy style tends be easier to read for long and in-depth docstrings.

The choice between styles is largely aesthetic, but the two styles should not be mixed. Choose one style for your project and be consistent with it. (source: sphinx documentation)

## Python modules
* For realistic projects you have several functions
   * putting them in different files/folders is usually a good option
   * a proper organization is crucial for code maintainability
* The first step is putting functions in different files creating so called Python modules
   * to use the functions in a module you have to `import` it
   * the module name (to be imported) corresponds to the file name
   * a module can host variables too
* When imported from another Python source file, the file name is treated as a namespace (remember the Zen? "*...Namespaces are one honking great idea -- let's do more of those!*")
   * when using the function, the name of the module must be prepended and separated by `.`
* You can import modules at the beginning of the calling unit, usually a good choice
   * or later, to make more readable where it is used
   * or even inside a function

In [23]:
%cat mymodule.py

nice_constant = 1.38064852e-23

def parabola(x):
    return x**2


In [24]:
import mymodule
print(mymodule.parabola(3))
print(mymodule.nice_constant)

9
1.38064852e-23


In [25]:
print("first do something")
import mymodule
print(mymodule.parabola(3))
print(mymodule.nice_constant)

first do something
9
1.38064852e-23


In [26]:
def myfunction():
    import mymodule
    print(mymodule.parabola(3))

myfunction()

9


## `import`ing modules
* It is possible to give an alias to the imported module
   * `import <module> as <module_alias>`
* It is also possible (and often recommended) to limit the `import` to a set of specific functions to be imported
   * `from <module> import <function>`
   * function can be used without prepending the module name
   * and you can import all module stuff using `*`
      * not a good practice, you do not have full control of what you are importing in the global namespace

## `import`ing modules - 2

* A module can be imported multiple times from the same code
   * under the hood, only one import is done 
   * no performance loss by importing it again
* Summarizing:
   * `import foo` 
   * `import foo as f`
   * `from foo import bar`
   * `from foo import * # <- do not do it, very bad`

In [27]:
import mymodule  # basic import
print(mymodule.parabola(3))

9


In [28]:
import mymodule as mm # import with renaming
print(mm.parabola(3))

9


In [29]:
from mymodule import parabola, nice_constant
# select functions to import
print(parabola(3))

9


In [30]:
from mymodule import * # select all functions to import
print(parabola(3))

9


## Organizing modules
* Module files can be stored in different folders 
   * in this case, you have many options to have Python looking into that folders
   * the simplest one is to specify the folder name in the import statement using `.` as separator
   * another option is setting PYTHONPATH environament variable
   * when you have a deep folder tree, the imports and calls can become deep too
* You have now many combinations of importing syntax!
   * absolute and relative imports are possible
   * much more to know, study it!

## Testing modules

* A very cool feature is the possibility to separately execute a module 
   * it is possible to specify actions to be executed only when the module is directly invoked (very useful for testing)
   * these actions are skipped when the module is imported
   * this can be achieved using a simple conditional
   
   `if __name__ == '__main__':`

In [31]:
def parabola(x):
    return x**2

if __name__ == '__main__':
    # test parabola function
    p = parabola(3.)
    print("check that p returns 9 in this case: ",p)

check that p returns 9 in this case:  9.0


## `math` module
* basic mathematical functions available in any Python distribution
* in CPython it is mostly a wrapper of C math library
* categories:
   * Number-theoretic and representation functions
   * Power and logarithmic functions
   * Trigonometric functions
   * Angular conversion
   * Hyperbolic functions
   * Special functions
   * Constants

In [32]:
import math
b = math.sin(3)
print(b)

0.1411200080598672


In [33]:
from math import ceil
d = ceil(1.2)
print(d)

2


## Hands-on 4.1


Find the solution of the equation 
$$f(x)=0$$ 
using Newton method
$$ x=x_0$$
$$ x_{n+1} = x_n-\frac{f(x_n)}{f'(x_n)}$$
Define the $f$ function and the $f'$ function as two separate Python functions. 
Test using $$f(x)=x^3$$ and starting from $$x_0=2$$


# Rights & Credits

These slides are CINECA 2018 and are released under the Attribution-NonCommercial-NoDerivs (CC BY-NC-ND) Creative Commons license, version 3.0. 

Uses not allowed by the above license need explicit, written permission from the copyright owner.

For more information see:
http://creativecommons.org/licenses/by-nc-nd/3.0/

Slides and examples were authored by:
**Sergio Orlandini**, **Francesco Salvadore**